cirq/_compat_test.py

```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
+_fake_b_deprecation_msg = [
+    'fake_b was used but is deprecated',
+    'Use cirq.testing._compat_test_data.module_a.module_b instead',
+] + _deprecation_origin
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_ops_deprecation_msg = [
+    'fake_ops was used but is deprecated',
+    'Use cirq.ops instead',
+] + _deprecation_origin
+
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_freezegun_deprecation_msg = [
+    'fake_freezegun was used but is deprecated',
+    'Use freezegun instead',
+] + _deprecation_origin
+
+
+def _trace_unhandled_exceptions(*args, **kwargs):
+    from multiprocessing import Queue
+
+    queue: Queue = kwargs['q']
+    func = kwargs['func']
+    del kwargs['q']
+    del kwargs['func']
+    try:
+        func(*args, **kwargs)
+        queue.put(None)
+    # coverage: ignore
+    except BaseException as ex:
+        # coverage: ignore
+        msg = str(ex)
+        queue.put((type(ex).__name__, msg, traceback.format_exc()))
+
+
+def subprocess_context(test_func):
+    """Ensures that sys.modules changes in subprocesses won't impact the parent process."""
+    import os
+
+    ctx = multiprocessing.get_context("spawn" if os.name == 'nt' else "fork")
+
+    exception = ctx.Queue()
+
+    def isolated_func(*args, **kwargs):
+        kwargs['q'] = exception
+        kwargs['func'] = test_func
+        p = ctx.Process(target=_trace_unhandled_exceptions, args=args, kwargs=kwargs)
+        p.start()
+        result = exception.get()
+        p.join()
+        if result:
+            # coverage: ignore
+            ex_type, msg, ex_trace = result
+            pytest.fail(f"{ex_type}: {msg}\n{ex_trace}")
+
+    return isolated_func
+
+
+@pytest.mark.parametrize(
+    'outdated_method,deprecation_messages',
+    [
+        (_from_parent_import_deprecated, [_fake_a_deprecation_msg]),
+        (_import_deprecated_assert_sub, [_fake_a_deprecation_msg]),
+        (_from_deprecated_import_sub, [_fake_a_deprecation_msg]),
+        (_import_deprecated_first_new_second, [_fake_a_deprecation_msg]),
+        (_import_new_first_deprecated_second, [_fake_a_deprecation_msg]),
+        (_import_multiple_deprecated, [_fake_a_deprecation_msg, _fake_b_deprecation_msg]),
+        (_new_module_in_different_parent, [_fake_ops_deprecation_msg]),
+        # ignore the frame requirement - as we are using find_spec from importlib, it
+        # is detected as an "internal" frame by warnings
+        (_find_spec_deprecated_multiple_times, [_fake_a_deprecation_msg[:-1]]),
+        (_import_parent_use_constant_from_deprecated, [_fake_a_deprecation_msg]),
+        (_import_deprecated_sub_use_constant, [_fake_a_deprecation_msg]),
+        (_import_deprecated_same_name_in_earlier_subtree, [_fake_a_deprecation_msg]),
+        (_import_top_level_deprecated, [_fake_freezegun_deprecation_msg]),
+        (_from_deprecated_import_sub_of_sub, [_fake_a_deprecation_msg]),
```

Comment:
It may be worth adding a test that just throws an error and is marked for pytest as expected to fail.  I think this would cover the above two `coverage: ignore`s.

Reply:
👍 great idea. Added.